app/livechat/server/api/lib/visitors.js

```diff
@@ -60,15 +62,36 @@ export async function findChatHistory({ userId, roomId, visitorId, pagination: {
 		skip: offset,
 		limit: count,
 	});
-
 	const total = await cursor.count();
-
 	const history = await cursor.toArray();
 
+	const resultArray = [];
+	if (text !== undefined) {
+		Meteor.runAsUser(userId, () => {
+			history.map(function(val) {
+				const roomId = val._id;
+				const count = 1;
+				const result = Meteor.call('messageSearch', text, roomId, count).message.docs;
+				if (result.length > 0) {
+					result.map(function(e) {
+						resultArray.push(e);
+					});
+				}
+			});
+		});
+	}
+	if (closedChatsOnly) {
```

Comment:
You need to fetch only closed rooms from the database, not test it after fetching the data from the DB.
Suggestion: Create a new method on the `LivechatRooms` model similar to the `findByVisitorId`, you can create a `LivechatRooms.findClosedByVisitorId` and then you can use the correct method according to the query parameter value:
```suggestion
	const options = { sort: sort || { ts: -1 }, skip: offset, limit: count };
	const cursor = closedChatsOnly  ? LivechatRooms.findClosedByVisitorId(visitorId, options)  : LivechatRooms.findByVisitorId(visitorId, options);
```

Reply:
done